File: historian-tree/src/main/java/net/yslibrary/historian/tree/HistorianTree.java
package net.yslibrary.historian.tree;

import net.yslibrary.historian.Historian;

import timber.log.Timber;

/**
 * Created by yshrsmz on 2017/01/21.
 */

public class HistorianTree extends Timber.Tree {

  private final Historian historian;

  private HistorianTree(Historian historian) {
    this.historian = historian;
  }

  public static HistorianTree with(Historian historian) {
    return new HistorianTree(historian);
  }

  @Override
  protected void log(int priority, String tag, String message, Throwable t) {
    historian.log(priority, message);
  }
}

File: historian-tree/src/main/java/net/yslibrary/historian/tree/HistorianTree.kt
package net.yslibrary.historian.tree

import net.yslibrary.historian.Historian
import timber.log.Timber

class HistorianTree private constructor(
    private val historian: Historian
) : Timber.Tree() {

    override fun log(priority: Int, tag: String?, message: String, t: Throwable?) {
        historian.log(priority, tag, message)
    }

    companion object {
        @JvmStatic
        fun with(historian: Historian): HistorianTree = HistorianTree(historian)
    }
}

/**
 * Extension function to create a Timber.Tree from a Historian instance
 */
fun Historian.asTimberTree(): Timber.Tree = HistorianTree.with(this)
